pkg/test/util/structpath/instance.go

```diff
@@ -17,82 +17,91 @@ package structpath
+
+	messagediff "gopkg.in/d4l3k/messagediff.v1"
+
 	"k8s.io/client-go/util/jsonpath"
 )
 
-var fixupNumericJSONComparison = regexp.MustCompile(`([=<>]+)\s*([0-9]+)\s*\)`)
-var fixupAttributeReference = regexp.MustCompile(`\[\s*'[^']+\s*'\s*]`)
+var (
+	fixupNumericJSONComparison = regexp.MustCompile(`([=<>]+)\s*([0-9]+)\s*\)`)
+	fixupAttributeReference    = regexp.MustCompile(`\[\s*'[^']+\s*'\s*]`)
+)
 
-type Structpath struct {
+type Instance struct {
 	structure interface{}
 	isJSON    bool
-	t         *testing.T
 }
 
-// Creates a new Structpath by marshaling the proto to JSON and then evaluating over that
+// New creates a structpath Instance by marshaling the proto to JSON and then evaluating over that
 // structure. This is the most generally useful form as serialization to JSON also automatically
 // converts proto.Any and proto.Struct to the serialized JSON forms which can then be evaluated
 // over. The downside is the loss of type fidelity for numeric types as JSON can only represent
 // floats.
-func AssertThatProto(t *testing.T, proto proto.Message) *Structpath {
+func New(proto proto.Message) (*Instance, error) {
 	if proto == nil {
-		t.Fatal("expected non-nil proto")
+		return nil, errors.New("expected non-nil proto")
+	}
+
+	parsed, err := protoToParsedJSON(proto)
+	if err != nil {
+		return nil, err
 	}
-	parsed := protoToParsedJSON(proto, t)
-	return &Structpath{parsed, true, t}
+
+	return &Instance{
+		structure: parsed,
+		isJSON:    true,
+	}, nil
 }
 
-func protoToParsedJSON(message proto.Message, t *testing.T) interface{} {
+func protoToParsedJSON(message proto.Message) (interface{}, error) {
 	// Convert proto to json and then parse into struct
 	jsonText, err := (&jsonpb.Marshaler{Indent: " "}).MarshalToString(message)
 	if err != nil {
-		t.Fatalf("failed to convert proto to JSON: %v", err)
+		return nil, fmt.Errorf("failed to convert proto to JSON: %v", err)
 	}
 	var parsed interface{}
 	err = json.Unmarshal([]byte(jsonText), &parsed)
 	if err != nil {
-		t.Fatalf("failed to parse into JSON struct: %v", err)
+		return nil, fmt.Errorf("failed to parse into JSON struct: %v", err)
 	}
-	return parsed
-}
-
-func (p *Structpath) ForTest(t *testing.T) *Structpath {
-	return &Structpath{p.structure, p.isJSON, t}
+	return parsed, nil
 }
 
-func (p *Structpath) Accept(path string, args ...interface{}) bool {
-	p.t.Helper()
+func (p *Instance) Accept(path string, args ...interface{}) error {
 	path = fmt.Sprintf(path, args...)
-	if p.findValue(path) == nil {
+	v, err := p.findValue(path)
+	if err != nil {
+		return err
+	}
+	if v == nil {
 		jsonText, _ := json.Marshal(p.structure)
-		p.t.Logf("Did not accept %v \n %v", path, jsonText)
-		return false
+		return fmt.Errorf("did not accept %v \n %v", path, jsonText)
 	}
-	return true
+	return nil
 }
 
-func (p *Structpath) Select(path string, args ...interface{}) *Structpath {
-	p.t.Helper()
+func (p *Instance) Select(path string, args ...interface{}) (*Instance, error) {
```

Comment:
If you have multi-valued returns, you can't do fluent composition. It is better to have a build-then-execute model, where the final execution can do the error return or execution. (e.g.)
```
structpath.ForProto(p)
.Select(...)
.Equals(...)
.Check() // Returns err
// or
.CheckOrFail(t) // fails test.

Reply:
I'm not sure that trying to mix fluent and non-fluent APIs is going to work very well. Fundamentally you're going to use one approach or the other .... very unlikely that you'd mix the two. Having to add `OrFail` to the method names makes the fluent-style rather ugly:

```golang
structPath.ForProto(p).
  SelectOrFail(...).
  EqualsOrFail(...).
  NotExistsOrFail(...).
  ExistsOrFail(...).
  ...
```
This seems overly verbose to me.  I'm really not sure what we gain by trying to put both styles into a single API.

Reply:
I am suggesting that you do the final check at the last function call (i.e. Check v.s. CheckOrFail) not at every step.

The alternative is to plumb it from the up top, but that requires two parallel implementations. Still viable IMO.

Reply:
> I am suggesting that you do the final check at the last function call (i.e. Check v.s. CheckOrFail) not at every step.

Wouldn't that require that we save the error (and possibly accumulate via multierror) along the way?

> The alternative is to plumb it from the up top, but that requires two parallel implementations. Still viable IMO.

I'm not exactly sure what you mean here by "plumb it from the up top", but two parallel implementations is the approach I'm taking here: one fluent api and one non-fluent.

Reply:
Sorry, for muddying the water, let me clean this up a bit. :)
I think what you're doing is fine. One downside of this is having to maintain a parallel implementation. Not a huge burden, but generally it is prone to issues with somebody adding a new method to one version, and not adding to the other one.

The general fluent pattern, in both cases is, first building a series of selections and constraints checks, and then executing them and harvesting the error. So, up till the last call, you would be just building a call graph. The final call will then execute the graph and collect the results (i.e. return error or fail). This allows you to retain a single type and a single set of builder methods. It also allows you to collect errors in all steps, not just the one that failed.

Reply:
@ozevren ah I see your point now. Makes sense.  Switched over to a single fluent-style api.